### PR TITLE
feat(route2): dynamic request aliasing

### LINKED ISSUE #387: Allow dynamically matching requests with a function
To allow highly flexible matching of which requests you want to stub, we could add the option to accept a function instead of a fixed method/URL. This would recieve the request object (or some wrapper around it). The motivation is that I have and SPA that uses a GraphQL backend where all requests have the same URL and I want to filter them by the query name which is sent as a parameter in the POST body.

```js
cy.route({
  matcher: (request) => !! request.body.match(//)
})
```

### PR DESCRIPTION
**Pull request by flotwig**
TR-369

- Closes #387

### User facing changelog

- Added the capability to dynamically add aliases for requests intercepted via `cy.route2`.

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue or this PR been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?  https://github.com/cypress-io/cypress-documentation/pull/3300
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/cypress/integration/commands/aliasing_spec.js [hunk 1]
```diff
@@ -497,19 +497,6 @@ describe('src/cy/commands/aliasing', () => {
         .get('input:first').as('firstInput')
         .get('@lastDiv')
       })
-
```
Comment (flotwig) [inline packages/driver/cypress/integration/commands/aliasing_spec.js:500]: this test was duplicated b/t waiting_spec and aliasing_spec

**Comment by bahmutov** [comment]
question: from the types it seems we can do `cy.route2({ url: '/users', alias: 'users' })` right? But there is no test there

**Comment by flotwig** [comment]
Nope, that's not a thing with this PR. That's the `RouteMatcher` type, which doesn't have an `alias` property, since it's supposed to match incoming HTTP requests.

**Review by bahmutov — APPROVED** [review]
super, it will make graphql matching much simpler

**Pull request merged**